### PR TITLE
Ensure NPE is always through when VirtualProcessor.onError(null) is invoked

### PR DESCRIPTION
**Pull request by jroper**
This fix is similar to #24749, fixing a spec violation bug that was introduced in #24722.

I've rearranged the code to ensure that hopefully it never regresses again, instead of having to check if the exception was null in every case, it now simply does it at the end, assuming it successfully reaches the end.

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala [hunk 1]
```diff
@@ -259,14 +256,17 @@ import scala.util.control.NonFatal
         case est @ Establishing(_, false, OptionVal.None) ⇒
           if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode($est).onError(${t.getMessage}), loop")
           if (!compareAndSet(est, est.copy(onErrorBuffered = OptionVal.Some(ex)))) rec(ex)
```
Comment (jroper) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: This is where the actual bug being fixed was - if establishing, and the error was successfully buffered, then an NPE wasn't thrown if error was null.

**Review by johanandren — COMMENTED** [review]
Thanks! Looks good. Could you add a test or two (I'm thinking `Subscriber` and `null` states should be easy to test) also, since it doesn't seem to be covered by the TCK?

**Review by johanandren — APPROVED** [review]
Oh wait, the actual regression was for `Establishing` and that would likely be impossible to reproduce repeatedly. NVM. LGTM!

**Comment by jroper** [comment]
Yeah, I don't think the case is practical to reproduce, though an interesting thing is, I'm pretty sure I upgraded to Akka Streams 2.5.12 (where the regression was introduced), and didn't encounter the issue, but then when I upgraded to 2.5.13, it now causes tests to fail more often than not. The thing that is reproducing this is the Akka Streams implementation of MicroProfile Reactive Streams Operators, which has a TCK that pretty much throws everything it can at the Reactive Streams TCK, for example here's the verification for the `limit` operator (equivalent to Akka Streams `drop`, called `limit` because it's following the JDK8 `java.util.streams` API naming):

https://github.com/eclipse/microprofile-reactive/blob/master/streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/LimitStageVerification.java

You can see at the bottom it provides a Reactive Streams TCK `IdentityProcessorVerification` against a `.limit(Long.MAX_VALUE)` processor.

In #24749 a test was introduced that exercised at least one of the paths in this code, and I'm sure there are other tests that exercise other parts - it's probably just luck that the existing tests haven't been tripped by this yet.

**Comment by jroper** [comment]
Ok, I just ran it against 2.5.12, it definitely exhibits in 2.5.12.

**Review by patriknw — APPROVED** [review]
LGTM, with an additional improvement of the debug lines

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala [hunk 2]
```diff
@@ -240,16 +240,13 @@ import scala.util.control.NonFatal
         case null ⇒
           if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode(null).onError(${t.getMessage}) -> ErrorPublisher")
```
[the remark this answers is not in the document]
Comment (jroper) [reply akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: Fixed.

**Comment by johanandren** [comment]
Fixes #25339
